### PR TITLE
Add hostname_replace plugin

### LINKED ISSUE #2609: Capability to rewrite hits url
**Is your feature request related to a problem? Please describe.**
Not related to a problem with `searx`, per se. I am looking to have `searx` rewrite the hits urls to another domain. 

**Describe the solution you'd like**

For example, if my results have a few urls from `www.reddit.com`, I'd like to have those rewitten to `old.reddit.com`.  If you could point me to where to start looking at, I'd be glad to send a PR if I am successful. My apologies if this already exists - if so, please guide how to accomplish this. 

In `settings.yml`, a section to include plain domain rewrites
<pre>
www.reddit.com: old.reddit.com
www1.domain.com: www3.domain.com
old.foo.org: new.foo.org
</pre>
**Describe alternatives you've considered**

Currently, I go to the address bar and perform manual rewrites. It would be great if `searx` can do the rewrites for me instead.  
**Additional context**

Sometimes, you'd want to access a specific endpoint for a results URI.

### PR DESCRIPTION
**Pull request by 3nprob**
## What does this PR do?

This is an alternative to #2638 , borrowing from the existing work done by @anmathew and @evuraan.

It supports replacing the domain/hostname (specifically, `result['parsed_url'].netloc`) based on regular expressions.

* Adds new plugin `hostname_replace`
* Configuration is done via "normal" settings
* Supports regular expressions

## Why is this change important?

Supports generic domain-replacements. Useful for redirecting sites to alternative frontends.

## How to test this PR locally?

To rewrite youtube links to use a specific invidious instance, add to `settings.yml`:
```
hostname_replace:
  '(.*\.)?youtube\.com$': 'invidiou.site'
  '(.*\.)?youtu\.be$': 'invidiou.site'
  '(.*\.)?youtube-noocookie\.com$': 'invidiou.site'
```
(please use on sites you manage or have permission from the owner, this is for illustration only)

## Related issues

Closes #2609, #1710

### CODE REVIEW

**Comment by 3nprob** [comment]
Room for improvement:

* Enable user configuration in frontend
  * it is not clear to me how this should be done without overcoupling the plugin to the backend
* Allow more complex rewrites: matching other parts of the URL, exempt-lists, etc
  * the current implementation should be good enough for major use-cases without being too complex

File: searx/plugins/hostname_replace.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+logger = logger.getChild(plugin_id)
+
+
+def on_result(request, search, result):
+    if parsed not in result:
+        return True
+    for (pattern, replacement) in replacements.items():
+        if pattern.search(result[parsed].netloc):
+            result[parsed] = result[parsed]._replace(netloc=pattern.sub(replacement, result[parsed].netloc))
```
Comment (return42) [inline searx/plugins/hostname_replace.py:25]: I did not know, that `._replace()` is an [documented method of named tuples](https://docs.python.org/3/library/collections.html#collections.somenamedtuple._replace).  Should we use it here / I don't now.

Comment (3nprob) [reply searx/plugins/hostname_replace.py:25]: Do you see any reason not to?
FWIW it's already in use by `tracker_url_remover`, and in `results.py`

**Comment by 3nprob** [comment]
Bringing over from https://github.com/searx/searx/pull/2638#issuecomment-813355895:
> @3nprob Do you like implement generic URL rewrites in #2724 (and rename it to generic_url_rewrite)?

Do you have any proposal of how that would look? There are several design directions to take and I figured by keeping this one smaller in scope there's still room for another more complete solution - the intention with this one is to server the use-case of replacing domains, not a generic rewrite DSL.

But - if you have a somewhat clear idea of how that would look I'm all ears!

**Comment by return42** [comment]
> There are several design directions to take and I figured by keeping this one smaller in scope there's still room for another more complete solution 

Before implementing several plugins, I guess it is better to have one for all.

> But - if you have a somewhat clear idea of how that would look I'm all ears!

Very simple: A regular expression substitution ([re.sub](https://docs.python.org/3/library/re.html#re.sub)) 

A list of expressions can be configured in `settings.yml` at engine level (_untested example_):

    - name : youtube
      shortcut : yt
      engine : youtube_noapi
      url_rewrite :
        '(http.*\.)?youtube\.com(.*)$': '\1invidiou.site\2'
        '(http.*\.)?youtu\.be(.*)$': '\1invidiou.site\2'
        '(http.*\.)?youtube-noocookie\.com(.*)$': '\1invidiou.site\2'

Bonus: A list of expressions can be configured in `settings.yml` at global level


    default_doi_resolver : 'oadoi.org'
    ...
    url_rewrite :
      '(http.*\.)?youtube\.com(.*)$': '\1invidiou.site\2'
      '(http.*\.)?youtu\.be(.*)$': '\1invidiou.site\2'
      '(http.*\.)?youtube-noocookie\.com(.*)$': '\1invidiou.site\2'

**Comment by 3nprob** [comment]
I started doing that actually, but the issue there is that the regexes quickly become pretty complex - and it's very easy to shoot yourself in the shoot and miss edge-cases.

Example for your pattern (amended for python regex syntax):

```
>>> re.sub('(http.*\.)?youtube\.com(.*)$', '\g<1>invidiou.site\g<2>', 'http://other.site/youtube.com/foo')
'http://other.site/invidiou.site/foo'
```

even if this one is addressed, covering for all edge-cases like these correctly yields complex expressions even for simple use-cases like domain replacement that I think are beyond what an average user should be expected to deal with

Performance is another one - once you do get a correct expression, you never know when you're hit by URLs that make it eat up all your RAM or hog up the CPU

Basically, covering the 80-90% use-cases and making it 10-20% as difficult to use 

This is from someone who has spent way too many hours of their life debugging URL-parsing regexes made by myself as well as other people ;)

**Comment by 3nprob** [comment]
@asciimoo thoughts of this one vs #2638? :)

**Comment by asciimoo** [comment]
I just realized that this aims to solve the same problem as #2638 . At first glance it looks simpler and more flexible. What disadvantages you see compared to #2638 ?

**Comment by 3nprob** [comment]
> I just realized that this aims to solve the same problem as #2638 . At first glance it looks simpler and more flexible. What disadvantages you see compared to #2638 ?

Honestly none really - meant to address the issues in that PR and with a more descriptive name.
I've been running it myself and it works fine for my use-cases (rewriting youtube/twitter/reddit/etc to alternative frontends)

**Comment by asciimoo** [comment]
Could you add some commented-out example to the settings.yml to demonstrate the usage?

**Comment by 3nprob** [comment]
> Could you add some commented-out example to the settings.yml to demonstrate the usage?
Will do!

**Comment by 3nprob** [comment]
@asciimoo updated

**Review by kvch — APPROVED** [review]
